src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java

```diff
@@ -1128,6 +1133,177 @@ void verifyJar(String jarName)
+                String entryName = locEntry.getName();
+                locEntries.add(entryName);
+
+                JarEntry cenEntry = jarFile.getJarEntry(entryName);
+                if (cenEntry == null) {
+                    crossChkWarnings.add(String.format(rb.getString(
+                            "entry.1.present.when.reading.jarinputstream." +
+                            "but.missing.via.jarfile"), entryName));
+                    continue;
+                }
+
+                try {
+                    readEntry(jis);
+                } catch (SecurityException e) {
+                    crossChkWarnings.add(String.format(rb.getString(
+                            "signature.verification.failed.on.entry.1." +
+                            "when.reading.via.jarinputstream"),
+                            entryName));
+                    continue;
+                }
+
+                try (InputStream cenInputStream = jarFile.getInputStream(cenEntry)) {
+                    if (cenInputStream == null) {
+                        crossChkWarnings.add(String.format(rb.getString(
+                                "entry.1.present.in.jarfile.but.unreadable"),
+                                entryName));
+                        continue;
+                    } else {
+                        try {
+                            readEntry(cenInputStream);
+                        } catch (SecurityException e) {
+                            crossChkWarnings.add(String.format(rb.getString(
+                                    "signature.verification.failed.on.entry.1." +
+                                    "when.reading.via.jarfile.inputstream"),
+                                    entryName));
+                            continue;
+                        }
+                    }
+                }
+
+                compareSigners(cenEntry, locEntry);
+            }
+
+            List<String> cenEntries;
+            cenEntries = jarFile.stream()
+                    .map(JarEntry::getName)
+                    .collect(Collectors.toList());
+
+            var cenEntries2 = cenEntries.getFirst().equals(JarFile.MANIFEST_NAME)
+                    ? cenEntries.subList(1, cenEntries.size()) : cenEntries;
+
+            Set<String> locEntrySet = new HashSet<>(locEntries);
+            Set<String> cenEntrySet = new HashSet<>(cenEntries2);
+
+            for (String cenEntry : cenEntrySet) {
+                if (!locEntrySet.contains(cenEntry)) {
+                    crossChkWarnings.add(String.format(rb.getString(
+                            "entry.1.present.when.reading.jarfile." +
+                            "but.missing.via.jarinputstream"), cenEntry));
+                }
+            }
+
+            if (!cenEntries2.equals(locEntries)) {
+                crossChkWarnings.add(rb.getString(
+                        "entries.mismatch.when.comparing.jarfile.and.jarinputstream"));
```

Comment:
Do we still need this warning? The meaning is not clear to me. Since we have already compared in both ways, does this only mean the orders are different?

Reply:
This step checks content and order. As the order does matter, I have this step to explicitly warn about ordering issue.

Reply:
But they are Sets now, so the order could have changed simply by adding them to the Set. I think this check can be removed.